### PR TITLE
GAWB-3612 Oauth with Fence

### PR DESCRIPTION
**Pull request by gpolumbo-broad**
\<your comments for this PR go here\>

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this **unless it's for a hotfix**. In that case, don't delete it!
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -51,13 +68,16 @@ def __init__(self):
...
     @endpoints.method(
         OAUTH_CODE_RESOURCE,
-        LinkInfoResponse,
+        AccessTokenResponse,
         path='fence/oauthcode',
         http_method='POST',
         name='fence/oauthcode')
     def oauthcode(self, request):
         user_info = self.auth.require_user_info(self.request_state)
-        return LinkInfoResponse(expires=datetime.now(), username=request.oauthcode)
```
Comment (dvoet) [inline main.py]: I think this should return LinkInfoResponse. Do we get any of this information from fence?

**Comment by dvoet** [comment]
I would like to see all the fence functionality in a different file than main with the actual calls to fence written in such a way as they can be mocked.

File: config.ini [hunk 2]
```diff
@@ -0,0 +1,6 @@
+[fence]
+CLIENT_ID=REPLACE_ME
+CLIENT_SECRET=REPLACE_ME
+AUTHZ_URI=https://qa.dcf.planx-pla.net/user/oauth2/authorize
+REDIRECT_URI=https://local.broadinstitute.org/fence-callback
+TOKEN_URL=https://qa.dcf.planx-pla.net/user/oauth2/token
```
Comment (dvoet) [inline config.ini]: why an ini file instead of environment variables in app.yml?

File: main.py [hunk 3]
```diff
@@ -43,21 +48,32 @@ class StatusResponse(messages.Message):
...
+        token_response = self.oauth_adapter.exchange_authz_code(request.oauthcode)
+        TokenStore.save(user_info.email, token_response)
```
Comment (dvoet) [inline main.py]: we don't want to use the email as a key, use user_info.id

File: bond_token.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+from google.appengine.ext import ndb
+import json
+
+
+class BondToken(ndb.Model):
+    token_dict_str = ndb.TextProperty()
+
+    REQUIRED_KEYS = ["access_token", "refresh_token", "token_type"]
```
Comment (dvoet) [inline bond_token.py]: We don't need to store access_token, it is only good for an hour. 

What is token_type and why do we need to store it?

I think we should also store the date it was created and the user name within the other system.

File: bond_token.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
+from google.appengine.ext import ndb
+import json
+
+
+class BondToken(ndb.Model):
```
Comment (dvoet) [inline bond_token.py]: This is an oauth refresh token. BondToken does not make much sense to me.

File: config.ini (hunk 2 above)
Comment (gpolumbo-broad) [reply config.ini]: This was done for testing.  Especially in intellij or I think even using nose, we can always read from the .ini, whereas app.yml isn't always loaded.

**Pull request merged**